### DISCUSSION TITLE
Make the start menu not delaying

### DISCUSSION THREAD

**Discussion by gailium119** [valinet/ExplorerPatcher#155] [General]
As said in previous issues, using the 22000.51 startmenuexperiencehost and Start_ShowClassicMode on 22000.65 or higher builds makes the start menu delay when it shuts down ,HOWEVER ,simply replacing twinui.dll and twinui.pcshell.dll with that in 22000.51
 makes it function properly.

**Comment by valinet** [comment]
Okay, that’s interesting. I wonder what changed between the 2 releases. By the way, the only difference between Start from 51 and Start from higher version regarding the old menu is that in 51 it checks the registry to see whether to enable the old menu, while in newer versions it check whether the Windows edition is Windows Team (PPI) or not.

**Comment by valinet** [comment]
Also, replacing those system files is not recommended. Patching them to work properly with the old Start is.

In fact, probably just one needs patching, the other is just required because the other just depends on it in some version specific way, I’d say.

**Reply by gailium119** [reply]
It's better than cold patching.though,since it doesn't break the signature

**Reply by valinet** [reply]
Hot patching is readily available from EP which injects `explorer`. Start menu communicates with that, does not load those 2 DLLs directly. The question is, what to patch...?

**Comment by valinet** [comment]
What about right clicking in Start? Does it work or it still crashes, with this workaround?

**Reply by gailium119** [reply]
it does,although by crashing it doesn't delay. I suppose another dll should be replaced

**Reply by valinet** [reply]
So does it crash or not? I still haven't understood.

**Reply by gailium119** [reply]
it crashes suddenly,not 'delaying' after crashing

**Reply by valinet** [reply]
So it still crashes. Yeah, another issue which needs fixing...

**Comment by valinet** [comment]
Do live tiles work? For me, they don't.

**Comment by gailium119** [comment]
Could you give me all the names of pdbs in the first build pf ExplorerPatcher?I assume that these dlls are the key

**Reply by valinet** [reply]
I don't understand what you want exactly.

**Reply by gailium119** [reply]
names like'Windows.ui.xaml'

**Reply by gailium119** [reply]
since they are the changed things in 22000.65

**Reply by valinet** [reply]
What names? From where?

**Reply by gailium119** [reply]
ExplorerPatcher downloads pdbs,right?

**Reply by valinet** [reply]
Yeah

**Reply by gailium119** [reply]
I need the names of them

**Reply by gailium119** [reply]
Not the latest build,since you've removed dependencies of some of them

**Reply by valinet** [reply]
Why is that relevant to this anyway? It's not EP that disables or cripples the Start menu...

**Reply by valinet** [reply]
Better than randomly checking files is to debug what happens when the menu is closed, what calls get made etc. You want to randomly replace system files until it works, or what is the strategy? Plus, replacing system files is not a solution, patching whatever got fucked up in the newer versions compared to the old versions is what should be done.

**Reply by gailium119** [reply]
It's not random since I use Beyondcompare to compare the difference of 22000.51 and 22000.65

**Reply by gailium119** [reply]
> Why is that relevant to this anyway? It's not EP that disables or cripples the Start menu...

the are the dependencies that may cause memory malfunctions

**Reply by valinet** [reply]
And you compare what? Why are the symbols downloaded by EP in some previous version relevant? That's what I don't understand.

**Reply by valinet** [reply]
> 
> 
> > Why is that relevant to this anyway? It's not EP that disables or cripples the Start menu...
> 
> the are the dependencies that may cause memory malfunctions

No, they're not, how did you reach this conclusion? EP does not have code specific for 51 or 65.

**Reply by gailium119** [reply]
> Better than randomly checking files is to debug what happens when the menu is closed, what calls get made etc. You want to randomly replace system files until it works, or what is the strategy? Plus, replacing system files is not a solution, patching whatever got fucked up in the newer versions compared to the old versions is what should be done.

Just knowing which file is the key can enable hot patching since the code can be reconstructed once the location of it is found.

**Reply by gailium119** [reply]
> And you compare what? Why are the symbols downloaded by EP in some previous version relevant? That's what I don't understand.

compare two 'bone' versions of 22000.51 and 22000.65 made by Tarsynear(find him on MDL forums)

**Reply by valinet** [reply]
I know that already, thanks for pointing it out. What I am saying is that you do not find the file among the list of symbols downloaded by EP in a previous version.

**Reply by valinet** [reply]
> 
> 
> > And you compare what? Why are the symbols downloaded by EP in some previous version relevant? That's what I don't understand.
> 
> compare two 'bone' versions of 22000.51 and 22000.65 made by Tarsynear(find him on MDL forums)

'bone' ??? Versions of what?

**Reply by gailium119** [reply]
> No, they're not, how did you reach this conclusion? EP does not have code specific for 51 or 65.

By experimenting with a virtual machine

**Reply by valinet** [reply]
And you determined that EP is the culprit why old Start does not work properly in new versions? C'mon, tf are you saying...

**Reply by gailium119** [reply]
> > > And you compare what? Why are the symbols downloaded by EP in some previous version relevant? That's what I don't understand.
> > 
> > 
> > compare two 'bone' versions of 22000.51 and 22000.65 made by Tarsynear(find him on MDL forums)
> 
> 'bone' ??? Versions of what?

a home-made esd archive that deletes almost everything including cortana,Hyper-V ,etc

**Reply by gailium119** [reply]
> And you determined that EP is the culprit why old Start does not work properly in new versions? C'mon, tf are you saying...

Because they are cummulative updates,not feature ones

**Reply by valinet** [reply]
Whatever, I don't understand a single thing that you are saying, I ask something you reply something else... you think EP is the reason old Start does not work in newer versions, that's simply nuts... test without EP, you will see old Start still behaves the same... I don't understand, you have an entire Git repo at your disposal, you ask me for some file list from an old version... like, go and get it yourself: https://github.com/valinet/ExplorerPatcher/blob/9bb7a4f154939b19332812b14795782b61eba6f5/ExplorerPatcher/dllmain.c. Although, it is useless, EP is not the culprit here, understand once and for all. 

> Because they are cummulative updates,not feature ones

What? Answer my question, not something random.

**Reply by gailium119** [reply]
what I need is the names instead of the pdbs ,since i want to determine the dependencies of StartMenuExperienceHost.exe

**Reply by valinet** [reply]
WHAT NAMES?????? What is a "dependency"?

**Reply by gailium119** [reply]
> WHAT NAMES?????? What is a "dependency"?

 names such as 'Windows.ui.xaml' ,'twinui.pcshell'
a dependency are another name of 'runtime'

**Reply by valinet** [reply]
... I am speechless... I don't understand a single thing that you are saying. I don't have a list of "dependencies" of old Start menu in EP's source code... whatever "dependency" may mean to you

**Reply by gailium119** [reply]
> ... I am speechless... I don't understand a single thing that you are saying. I don't have a list of "dependencies" of old Start menu in EP's source code... whatever "dependency" may mean to you

Say it easy as 'the start menu can't open if these dlls don't exist '

**Reply by valinet** [reply]
And how would I know that? And obtaining a complete list is not easy at all, either...

For a start, if you want to find out the libraries a program imports functions from, you can use "dumpbin /dependents". To view the imported function names, use "dumpbin /imports".

But that is not enough. Function importing is not the single mechanism via which external code can be executed. That could still be done at runtime using `GetProcAddress`, or via other mechanisms, like COM, for example. That you figure out by studying the disassembly or the source code, if you have access to that. It takes a while to figure it out, that's what this entire EP project is about.

Anyway, here is "dumpbin /dependents "C:\Windows\SystemApps\Microsoft.Windows.StartMenuExperienceHost_cw5n1h2txyewy\StartMenuExperienceHost.exe":

```
Microsoft (R) COFF/PE Dumper Version 14.29.30136.0
Copyright (C) Microsoft Corporation.  All rights reserved.


Dump of file C:\Windows\SystemApps\Microsoft.Windows.StartMenuExperienceHost_cw5n1h2txyewy\StartMenuExperienceHost.exe

File Type: EXECUTABLE IMAGE

  Image has the following dependencies:

    api-ms-win-eventing-provider-l1-1-0.dll
    api-ms-win-core-debug-l1-1-0.dll
    api-ms-win-core-processthreads-l1-1-0.dll
    api-ms-win-core-localization-l1-2-0.dll
    api-ms-win-core-heap-l1-1-0.dll
    api-ms-win-core-libraryloader-l1-2-0.dll
    api-ms-win-core-threadpool-l1-2-0.dll
    api-ms-win-core-synch-l1-1-0.dll
    api-ms-win-core-errorhandling-l1-1-0.dll
    api-ms-win-core-registry-l1-1-0.dll
    api-ms-win-core-sysinfo-l1-2-0.dll
    api-ms-win-core-synch-l1-2-0.dll
    api-ms-win-core-winrt-string-l1-1-0.dll
    api-ms-win-core-handle-l1-1-0.dll
    api-ms-win-core-com-l1-1-0.dll
    wincorlib.DLL
    api-ms-win-crt-runtime-l1-1-0.dll
    api-ms-win-crt-string-l1-1-0.dll
    api-ms-win-crt-locale-l1-1-0.dll
    api-ms-win-crt-private-l1-1-0.dll
    api-ms-win-core-winrt-error-l1-1-0.dll
    api-ms-win-core-winrt-error-l1-1-1.dll
    ntdll.dll
    api-ms-win-core-util-l1-1-0.dll
    api-ms-win-core-string-l1-1-0.dll
    api-ms-win-core-processthreads-l1-1-1.dll
    api-ms-win-core-profile-l1-1-0.dll
    api-ms-win-core-sysinfo-l1-1-0.dll
    api-ms-win-core-interlocked-l1-1-0.dll

  Summary

        7000 .data
        D000 .pdata
       3A000 .rdata
        3000 .reloc
       A4000 .text
```

Dumpbin comes with Build Tools for Visual Studio or Visual Studio. 

See, it is much more complicated. You won't find those "api-ms-..." files anywhere. They are not real files, they just give the loader information regarding a mechanism (Windows API sets: https://docs.microsoft.com/en-us/windows/win32/apiindex/windows-apisets) via which the loader determines whether those should be redirected at runtime to either a DLL implementing that functionality, or some stub DLL with minimal prototypes where those functions fail, depending on whether that particular Windows version has that functionality. Those are resolved only at runtime by the loader, the executable file does not contain information about where the actual code of that function prototype is located.

Good luck!

**Reply by gailium119** [reply]
the dependencies can be checked by process explorer, a utility of sysinternals(https://docs.microsoft.com/en-us/sysinternals); however,i've replaced almost system files that  replacing the ones not replaced will cause the system unbootable, but the start menu still crashes when right-clicking. I suggest a registry difference causing an error of not finding components because of the version. So I've uploaded the files and hope that you have time to investigate it.

**Reply by gailium119** [reply]
https://1drv.ms/u/s!Av-jwLLlKnISgTIkNyKLEdi-IkPM

**Reply by valinet** [reply]
I barely have time to do anything. The way to debug this is to hook specific points in StartDocked and also watch generally with a debugger. Replacing system files is not a proper option, because it break at each update. It can largely give a clue over where the problem is, but you have to look more deeper to find it specifically. And again, the only proper solution is injecting the process and hotpatching it (at runtime).

**Reply by gailium119** [reply]
How to debug the startmenuexperiencehost.exe to find why it crashes?

**Comment by gailium119** [comment]
Abort mission.Found that although the main version(22000) cannot be changed by registry,changing the ubr registry key(HKLM\software\microsoft\windows nt\currentversion\UBR) to 65 or higher tricks powershell and a package like the test-only package of wsa
[Downloads.zip](https://github.com/valinet/ExplorerPatcher/files/7380898/Downloads.zip)
which can be installed only on 22000.63 or higher to install it.

**Reply by valinet** [reply]
What is wsa?

**Reply by Gaurav-Original-ClassicShellTester** [reply]
> What is wsa?

He probably means Windows Subsystem for Android

**Reply by gailium119** [reply]
right

**Comment by gailium119** [comment]
Just found,wsa cannot be installed on a version less than 2200.120(although by a msix repack it can).
So maybe I still need to work on restoring the old start menu.I guess there is something to do with windows feature experiernce pack.

**Comment by valinet** [comment]
Attach using a debugger, like Windbg. It will show a stack trace when it crashes. That’s a good starting point.

Valentin-Gabriel Radu

**Reply by gailium119** [reply]
I tried and the error includes a class not registered and a file not found;
however,i can't determine which class and file. Cam I use ida instead of windbg? It is more efficient to me.
